### ISSUE TITLE
catalog configuration annotation

### ISSUE THREAD

**Issue by jrchudy** [informatics-isi-edu/chaise#1704]
We discussed a few different config patterns to use for the catalog annotation.

## Support templating in navbarMenu (#1709)
- Should we change the name of the property e.g.  `navbarMenu` v.s. `navbarMenuTemplate`? 
- If both exist, navbarMenuTemplate overwrite?  
- The URL can be static or version specific. Users have the option of using `$catalog.snapshot` in the template to direct users to specific version based on the current URL.   

## Order of setting chaise configuration
In short, apply the most generic config first, then overwrite with more and more specific config. The order should be:
1. preset chaise-config defaults
2. chaise-config.js 
3. catalog annotation in the live catalog
4. catalog annotation in the specific catalog snapshot, if exists  

Note: In 2-4, loop over each rule from top to bottom 
    - if match, apply all configuration properties
    - any properties defined in current match will override any previously set properties from a previous configuration settings.

# Option 1
- Generally follow the existing chaise-config syntax and extend the config with rules that can overwrite some values for different server name.
- Use the same syntax in `"tag:isrd.isi.edu,2019:chaise-config":{ // chaise-config syntax }`
- Follow the order of rule-overwriting according to the section above. 
- The advantages are: 1) it is consistent with existing chaise-config; 2) it allows an explicit section to describe generic config parameters; 3) the document is self-contained and can be used for specific snapshots; 4) the navbar can be written per host for static site, and/or the same config can be moved to catalog annotation. 

```
  "configRules": [
     { 
       "host": <hostname> | [<hostname>, ...],
       "config": {
          // configuration properties here
        } 
    },
    {}
  ]
```
Example:
```
{
   // generic chaise-config
    catalog: 2,   // default catalog
    headTitle: "RBK/GUDMAP",
    navbarBrand: "/resources/",
    navbarMenu: {
   },
    "configRules": {
     { 
       // only support host as part of the configuration at the moment
       "host": ["www.rebuildingakidney.org", "staging.rebuildingakidney.org", "dev.rebuildingakidney.org"],
       "config": {
        "headTitle": "RBK/GUDMAP",
        "navbarBrand": "/resources/",
        "navbarMenu" : { 
           // rbk specific navbar
          { "name": "Protocols", "url": "/chaise/recordset/{{{$catalog.snahsot}}}/Protocol:Protocol/Consortium="RBK'"}
        } 
   },
   {
       "host": ["www.gudmap.org", "staging.gudmap.org", "dev.gudmap.org"],
       "config": {
        "headTitle": "GUDMAP/RBK",
        "navbarBrand": "/",
        "navbarMenu" : { 
          // GUDMAP specific navbar
          { "name": "Protocols", "url": "/chaise/recordset/{{{$catalog.snahsot}}}/Protocol:Protocol/Consortium="GUDMAP'"}
        } 
    }
```
NOTE: for snapshot navbar to work properly, we will need to amend the history with proper chaise-config catalog annotation through the history API `POST /ermrest/history/<from>,<to>/annotation/<catalog RID>`

# Option 2
- Use catalog annotation to configure chaise-config rules. 
- Follow the order of rule-overwriting according to the section above. 
- Issue: to make the annotation makes sense by itself (e.g. needed for history), a generic rule that apply to all config needs to be written e.g. an object containing only "config" object .

```
"tag:isrd.isi.edu,2019:chaise-config": [
    {
      "host": <hostname> | [<hostname>, ...],
      "catalog": <catalog-id> | [<catalog-id>, ...],
      "version": {
          min: <value> | null (same as not defined), 
          max: <value> | true | null (same as not defined) 
       },  // version range
      "config": {
          configuration properties here
      }, 
      {
        ...
      }, ...
]
```
Each of the options (host, catalog, and version) would be optional. **HT: when they are missing, the rule apply to all hosts, all catalogs, all versions.** They are just a way to be specific as to when the rule should be selected to use that configuration. 

### version property
 - version is not defined | `null`, it applies to all catalogs (live and versioned)
 - version is an object
   - min
     - not defined | `null`, all versions before defined `max` will be used
     - <version>, starting at this version (and including) until defined `max`
   - max 
     - not defined | `null`, all versions after defined `min` including current/live catalog
     - `true`, all versions after defined `min` **NOT** including current/live catalog
     - <version>, before this version (and including)
   - min is `null` and max is `null`, `{ "min": null, "max": null }`
     - same as when version is not defined | `null`, applies all catalogs (live and versioned)
   - version is empty object (`{}`)
     - same as above case if min and max were set to `null`, applies to all catalogs (live and versioned)

There was discussion as to which properties should go into the "config" block. It makes sense to potentially include all properties that are configurable in the chaise config.

For example:
```
"tag:isrd.isi.edu,2019:chaise-config": [
  {
    "config": {
      "editRecord": true,
      "deleteRecord": false,
      "navbarMenu": {
        ...
      }
    }
  }, {
    "catalog": 1,
    "config": {
      "navbarMenu": {
        ...
      }
    }
  }
]
```

In the above example, we have a specific configuration for "everything" and another more specific configuration for catalog 1.

### Chosen Option:
We look at the configurations in the order they are defined in the array. Each rule will be checked in order to verify if it matches the current conditions. If it does match, all properties defined will be applied. The next configuration will be checked and follow the same strategy of checking if the conditions are met. If a match is found, all options will be applied (NOTE: this will override any properties defined in an early matching configuration.

This method would allow for inheritance between configurations and not force annotators to have duplicate configuration options between configuration rules.

Following this method, if we are looking at catalog 1, given the above configuration, we would look at the general configuration first (first configuration in the list) and set the `editRecord`, `deleteRecord`, and `navbarMenu` properties. Then we would look at the second rule to then apply the `navbarMenu` property again which would overwrite the previously defined one.

## Interaction with `chaise-config.js` and Defaults
This leaves one more point of discussion, this has only been about how the annotation would interact with itself. But we still have to consider what it means to have the annotation defined alongside a chaise-config document in the server.

The chaise-config annotation and the `chaise-config.js` document will follow the same object structure that is decided on so that the same rules are applied in the same way.

~~Regardless of the decision made for how the annotation will behave, either inheritance between rules or match the most specific rule ONLY, we have to consider how the annotation will interact with an existing chaise-config that may include any of the same set of properties.~~

~~We can use the same 2 approaches as above, either use inheritance or an all or nothing approach. If we use inheritance, we have to define what that inheritance means. ~~
~~ 1. After going through the rules that are decided on for the annotation, should we then fallback to the chaise config document (if no rule matches in the annotation)?    - this seems like the case, we would want to support the annotation config, fallback on the chaise-config document, and then finally fallback on the chaise-config defaults (global defaults) that are defined in the [documentation](https://github.com/informatics-isi-edu/chaise/blob/master/docs/user-docs/chaise-config.md).~~
~~2. If we do match a rule in the annotation, should we still look at the `chaise-config.js` document on the server to apply more properties that weren't specified in the annotation (global properties for the server)?~~

**Comment by carlkesselman** [comment]
I would argue in favor that the config block should allow anything that could be in chaise config.  The only exception I would make is that if there was an option that really only applied to the server and not a catalog.
In general, I would like to see a situation where I don’t have to ssh into the server at all but could do everything via ERMrest.

Carl

----------------------------------------------------------
Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute
Viterbi School of Engineering

Professor, Preventive Medicine
Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA  90292-6695
Phone:  +1 (310) 448-9338
Email:  carl@isi.edu<mailto:carl@isi.edu>
Web: http://www.isi.edu/~carl








On Jan 25, 2019, at 3:42 PM, Josh Chudy <notifications@github.com<mailto:notifications@github.com>> wrote:


We discussed a few different config patterns to use for the catalog annotation.

Rule Matching config

Case with rule matching:

"tag:isrd.isi.edu<http://isrd.isi.edu>,2019:catalog-chaise-config": {
  "config-rules": [
    {
      "host": <hostname>,
      "catalog": <catalog-id>,
      "version": <true, false, version range>,
               // true means apply to all version,
               // false means don't apply to any versions
               // range means which versions to apply for
      "config": {
          configuration properties here
      },
      {
        ...
      }, ...
  ]
}


Each of the options (host, catalog, and version) would be optional. They are just a way to be specific as to when the rule should be selected to use that configuration.

There was discussion as to which properties should go into the "config" block. It makes sense to potentially include all properties that are configurable in the chaise config. If that's the case, we have to decide whether only one config block will ever apply or if there should be a hierarchy of how the rules would apply.

For example:

"config-rules": [
  {
    "catalog": 1,
    "config": {
      "navbarMenu": {
        ...
      }
    }
  }, {
    "config": {
      "editRecord": true,
      "deleteRecord": false,
      "navbarMenu": {
        ...
      }
    }
  }
]


In the above example, we have a specific configuration for catalog 1, and another configuration for "everything else".

Option 1:

Should we look at all configurations and apply the most specific config options first, then look at the next most specific configuration that applies, and apply those config options that haven't been applied yet? This method would allow for inheritance between configurations and not force annotators to have duplicate configuration options between configuration rules.

Following this method, if we are looking at catalog 1, given the above configuration, we would use the navbarMenu configuration from the first rule and then look at the second rule to then apply the editRecord and deleteRecord configuration.

Option 2

The alternative would be to find the most specific configuration rule that best matches, and use only that defined configuration. So if we were looking at catalog 1, given the above configuration, ONLY the navbarMenu configuration option would be applied.

Overrides config

**Comment by jrchudy** [comment]
The above issue text has been updated based on more discussion that we had during our morning meeting today. To address some of @carlkesselman 's questions:

`the config block should allow anything that could be in chaise config`
 - we have decided that all chaise config properties will be configurable in the annotation

`if there was an option that really only applied to the server and not a catalog`
 - in the case of the default `catalogId` this makes perfect sense, it would be redundant to have that as a catalog annotation option (and possibly some other options, I haven't looked too deeply into it yet)

`I don’t have to ssh into the server at all`
 - Part of my answer to the second statement applies here. We need to figure out which properties should be "server configurable" only, so the need to ssh and edit something won't occur. This doesn't take away from the fact that some properties *may* still be configured at the server level as well as at the catalog level
